File: Solutions/1547. Minimum Cost to Cut a Stick/1547. Minimum Cost to Cut a Stick.py
# Top down
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        dp=[[0]*102 for _ in range(102)]
        def dfs(cuts,i,j):
            if j-i<=1:
                return 0
            if not dp[i][j]:
                dp[i][j]=sys.maxsize
                for k in range(i+1,j):
                    dp[i][j]=min(dp[i][j],cuts[j]-cuts[i]+dfs(cuts,i,k)+dfs(cuts,k,j))
            return dp[i][j]
        cuts.append(0)
        cuts.append(n)
        cuts.sort()
        return dfs(cuts,0,len(cuts)-1)

# Bottom Up
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        cuts.append(0)
        cuts.append(n)
        cuts.sort()
        size=len(cuts)
        dp=[[0]*size for _ in range(size)]
        for i in range(size-1,-1,-1):
            for j in range(i+1,size):
                for k in range(i+1,j):
                    dp[i][j]=min(sys.maxsize if dp[i][j]==0 else dp[i][j],cuts[j]-cuts[i]+dp[i][k]+dp[k][j])
        
        return dp[0][size-1]

File: Solutions/1547. Minimum Cost to Cut a Stick/1547. Minimum Cost to Cut a Stick.go
//Top Down

import (
    "math"
    "sort"
)
func minCost(n int, cuts []int) int {
    dp:=make([][]int,102)
    for i,_:=range(dp){dp[i]=make([]int,102)}
    cuts=append(cuts,0)
    cuts=append(cuts,n)
    sort.Ints(cuts)
    return dfs(&dp,&cuts,0,len(cuts)-1)
}

func dfs(dp *[][]int,cuts *[]int,i,j int)int{
    if j-i<=1{
        return 0
    }
    if (*dp)[i][j]==0{
        (*dp)[i][j]=math.MaxInt32
        for k:=i+1;k<j;k+=1{
            (*dp)[i][j]=min((*dp)[i][j],(*cuts)[j]-(*cuts)[i]+dfs(dp,cuts,i,k)+dfs(dp,cuts,k,j))
        }
    }
    return (*dp)[i][j]
}

func min(i,j int)int{
    if i<j{
        return i
    }
    return j
}


//Bottom Up

import (
    "math"
    "sort"
)
func minCost(n int, cuts []int) int {
    cuts=append(cuts,0)
    cuts=append(cuts,n)
    sort.Ints(cuts)
    size:=len(cuts)
    dp:=make([][]int,size)
    for i,_:=range(dp){dp[i]=make([]int,size)}
    for i:=size-1;i>=0;i-=1{
        for j:=i+1;j<size;j+=1{
            for k:=i+1;k<j;k+=1{
                if dp[i][j]==0{
                    dp[i][j]=math.MaxInt32
                }
                dp[i][j]=min(dp[i][j],dp[i][k]+dp[k][j]+cuts[j]-cuts[i])
            }
        }
    }
    return dp[0][len(cuts)-1]
}

func min(i,j int)int{
    if i<j{
        return i
    }
    return j
}